test/test_linearizer.py

```diff
@@ -202,7 +202,7 @@ def test_early_end_local(self):
     k.hand_coded_optimizations()
     k.linearize()
     self.assertEqual(len(endifs:=[x for x in k.uops if x.uop is UOps.ENDIF]), len(ifs:=[x for x in k.uops if x.uop is UOps.IF]))
-    self.assertEqual(len(barriers:=[x for x in k.uops if x.uop is UOps.BARRIER]), 3)
+    self.assertEqual(len(barriers:=[x for x in k.uops if x.uop is UOps.BARRIER]), 4)
```

Comment:
why does GPU need 4 barriers but other backends don't?

Reply:
oh I see how this could happen, you have a warp size of 32 and issues like this can hide.
I think the best way to catch these is have tests with local size of > 32, can you add some with larger lidx?

Reply:
I'm fine with changing UOps.BARRIER vins in this case. Correctness is more important.

Reply:
sweet, I'll play with it and draw something up